File: sobelAndLaplaceColor.py
import numpy as np
import cv2
import matplotlib.pyplot as plt

def sobelAndLaplaceColor(input_image, thresholdValue=100, process_mode='RGB'):
    # Read the image
    I = cv2.imread(input_image)

    # Check if the image is grayscale
    if len(I.shape) == 2 or I.shape[2] == 1:
        I = cv2.cvtColor(I, cv2.COLOR_GRAY2BGR)
    
    # Convert to double for calculations
    input_image = I.astype(np.float64)

    # Sobel Operator Masks
    Mx = np.array([[-1, 0, 1], [-2, 0, 2], [-1, 0, 1]])  # Horizontal mask
    My = np.array([[-1, -2, -1], [0, 0, 0], [1, 2, 1]])  # Vertical mask

    # Laplace Operator Mask
    laplace_kernel = np.array([[0, -1, 0], [-1, 4, -1], [0, -1, 0]])

    # Convert to grayscale for luminosity processing
    if process_mode == 'LUM':
        I_gray = cv2.cvtColor(I, cv2.COLOR_BGR2GRAY)
        I = np.stack([I_gray] * 3, axis=-1)  # Treat grayscale as 3 identical channels

    # Pre-allocate matrices for Sobel and Laplace results
    sobel_image = np.zeros_like(I, dtype=np.uint8)
    laplace_image = np.zeros_like(I, dtype=np.uint8)

    # Edge Detection Process for Each Channel
    for channel in range(3):
        # Extract the channel
        channel_data = input_image[:, :, channel]

        # Sobel Filtering
        Gx = cv2.filter2D(channel_data, -1, Mx)
        Gy = cv2.filter2D(channel_data, -1, My)
        sobel_filtered = np.sqrt(Gx**2 + Gy**2)

        # Laplace Filtering
        laplace_filtered = cv2.filter2D(channel_data, -1, laplace_kernel)

        # Normalize results
        sobel_image[:, :, channel] = cv2.convertScaleAbs(sobel_filtered)
        laplace_image[:, :, channel] = cv2.convertScaleAbs(laplace_filtered)

    # Apply threshold for Laplacian
    laplace_image[laplace_image < thresholdValue] = 0

    # Display results
    plt.figure(figsize=(12, 8))
    plt.subplot(1, 3, 1)
    plt.imshow(cv2.cvtColor(I, cv2.COLOR_BGR2RGB))
    plt.title('Input Image')
    plt.axis('off')
    
    plt.subplot(1, 3, 2)
    plt.imshow(cv2.cvtColor(sobel_image, cv2.COLOR_BGR2RGB))
    plt.title('Sobel Edge Detection')
    plt.axis('off')
    
    plt.subplot(1, 3, 3)
    plt.imshow(cv2.cvtColor(laplace_image, cv2.COLOR_BGR2RGB))
    plt.title('Laplace Edge Detection')
    plt.axis('off')
    
    plt.show()
    
    return sobel_image, laplace_image

if __name__ == "__main__":
    # Path to your image
    image_path = '/Users/Lorna/Pepper.tif'  # Change this to your image path
    threshold = 100  # Threshold for Laplacian
    process_mode = 'RGB'  # Options: 'RGB' or 'LUM'

    # Apply Sobel and Laplacian edge detection
    sobel_image, laplace_image = sobelAndLaplaceColor(image_path, threshold, process_mode)

File: sobelAndLaplaceColor.m
function [sobel_image, laplace_image] = sobelAndLaplaceColor(input_image, thresholdValue)
    % Read and validate the input image
    I = imread(input_image);

    % Check if the image is colored (RGB)
    if size(I, 3) ~= 3
        error('Input must be a colored image (RGB).');
    end

    % Convert to double for calculations
    input_image = double(I);

    % Sobel Operator Masks
    Mx = [-1 0 1; -2 0 2; -1 0 1]; % Horizontal mask
    My = [-1 -2 -1; 0 0 0; 1 2 1]; % Vertical mask

    % Laplace Operator Mask
    laplace_kernel = [0 -1 0; -1 4 -1; 0 -1 0];

    % Pre-allocate matrices for Sobel and Laplace results
    sobel_image = zeros(size(input_image), 'uint8');
    laplace_image = zeros(size(input_image), 'uint8');

    % Edge Detection Process for Each Channel
    for channel = 1:3
        % Extract the channel
        channel_data = input_image(:, :, channel);

        % Get image dimensions
        [rows, cols] = size(channel_data);

        % Pre-allocate the filtered images
        sobel_filtered = zeros(rows, cols);
        laplace_filtered = zeros(rows, cols);

        % Sobel Filtering
        for i = 2:rows-1
            for j = 2:cols-1
                % Extract the 3x3 region
                region = channel_data(i-1:i+1, j-1:j+1);

                % Gradient approximations
                Gx = sum(sum(Mx .* region));
                Gy = sum(sum(My .* region));

                % Calculate the magnitude
                sobel_filtered(i, j) = sqrt(Gx.^2 + Gy.^2);
            end
        end

        % Laplace Filtering
        for i = 2:rows-1
            for j = 2:cols-1
                % Extract the 3x3 region
                region = channel_data(i-1:i+1, j-1:j+1);

                % Laplacian approximation
                laplace_filtered(i, j) = sum(sum(laplace_kernel .* region));
            end
        end

        % Normalize results
        sobel_image(:, :, channel) = uint8(255 * mat2gray(sobel_filtered));
        laplace_image(:, :, channel) = uint8(255 * mat2gray(laplace_filtered));
    end

    % Apply threshold for Laplacian if given
    if nargin < 2
        thresholdValue = 100; % Default threshold
    end
    laplace_image(laplace_image < thresholdValue) = 0;

    % Convert logical image back to uint8
    laplace_image = uint8(laplace_image);

    % Display results
    figure, imshow(I); title('Input Image');
    figure, imshow(sobel_image); title('Sobel Edge Detection');
    figure, imshow(laplace_image); title('Laplace Edge Detection');
end
